### PR TITLE
Large strings support in cudf::merge

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Enable large strings support in `cudf::merge`. 
Simplifies the strings specialization to use the gather-based strings factory function which is already optimized for long strings and is now appropriately enabled for large strings.
Also moved source from `include/cudf/strings/detail/merge.cuh` to `src/strings/merge/merge.cu` file since the template implemenation was not actually required.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by robertmaynard — APPROVED** [review]
Approving CMake changes

File: cpp/src/strings/merge/merge.cu [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/detail/merge.hpp>
+#include <cudf/strings/detail/strings_column_factories.cuh>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/functional>
```
Comment (bdice) [inline cpp/src/strings/merge/merge.cu:27]: Is this header used? Looks like the `__device__` lambda didn't need to have its return type proclaimed, which surprises me a bit. Most `thrust::transform` calls do need a return type to be proclaimed. You would see compile errors if something was wrong.

Maybe check for other unused headers while you're at it.

File: cpp/include/cudf/strings/detail/strings_children.cuh [hunk 2]
```diff
@@ -163,22 +163,23 @@ std::pair<std::unique_ptr<column>, int64_t> make_offsets_child_column(
...
-  // }
-
-  return std::pair(std::move(offsets_column), total_elements);
+  auto const threshold = get_offset64_threshold();
+  if (!is_large_strings_enabled()) {
+    CUDF_EXPECTS(
+      total_bytes < threshold, "Size of output exceeds the column size limit", std::overflow_error);
+  }
+  if (total_bytes >= get_offset64_threshold()) {
+    // recompute as int64 offsets when above the threshold
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/strings_children.cuh:175]: I don't really expect large strings to be very common. At least not in the near to mid future.
I'd rather force any performance hit into the int64 path right now if possible.

**Review by bdice — APPROVED** [review]

File: cpp/tests/merge/merge_string_test.cpp [hunk 3]
```diff
@@ -411,3 +411,48 @@ TYPED_TEST(MergeStringTest, Merge2StringKeyNullColumns)
   CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_column_view2, output_column_view2);
   CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_column_view3, output_column_view3);
 }
+
+class MergeLargeStringsTest : public cudf::test::BaseFixture {};
+
+TEST_F(MergeLargeStringsTest, MergeLargeStrings)
+{
```
Comment (nvdbaranec) [inline cpp/tests/merge/merge_string_test.cpp:418]: Super small nit:  maybe a negative test verifying 32 bit for small output columns?

Comment (davidwendt) [reply cpp/tests/merge/merge_string_test.cpp:418]: Wouldn't that be all the other tests in this file? 
Or do you mean adding a check to one of those tests that the offsets are 32-bit?

File: cpp/include/cudf/strings/detail/merge.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <cudf/column/column.hpp>
+#include <cudf/detail/merge.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/merge.hpp:26]: Nit (even smaller than @nvdbaranec nit): We can write such nested namespace shorter as `cudf::strings::detail`.

File: cpp/tests/merge/merge_string_test.cpp (hunk 3 above)
Comment (nvdbaranec) [reply cpp/tests/merge/merge_string_test.cpp:418]: > Wouldn't that be all the other tests in this file? Or do you mean adding a check to one of those tests that the offsets are 32-bit?

Right.  Or just an extension in this one that uses a smaller multiplier to produce a < 2 GB chars column.

File: cpp/include/cudf/strings/detail/merge.hpp (hunk 4 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/merge.hpp]: Done

File: cpp/include/cudf/strings/detail/merge.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf ::strings ::detail {
+/**
+ * @brief Merges two strings columns
+ *
+ * @param lhs First column
+ * @param rhs Second column
+ * @param row_order Indices for each column
```
Comment (vuule) [inline cpp/include/cudf/strings/detail/merge.hpp:30]: I noticed that `Caller must set the validity mask in the output column.` has been removed. Is there really a change in behavior?

Comment (davidwendt) [reply cpp/include/cudf/strings/detail/merge.hpp:30]: Yes, the validity mask is computed in the `make_strings_column` utility and so the caller does not need to compute the mask anymore. 
It has been removed here:
https://github.com/rapidsai/cudf/pull/15374/files#diff-bea7124eddd4e287917be21bd0a0b7cb946816ad08c2a77943177a54569d87d0L443-L447

Comment (vuule) [reply cpp/include/cudf/strings/detail/merge.hpp:30]: yup, saw that as well. Great stuff.

Comment (davidwendt) [reply cpp/include/cudf/strings/detail/merge.hpp:30]: Good catch too. You had to look at the deleted file to find that.

**Review by nvdbaranec — APPROVED** [review]

**Review by vuule — COMMENTED** [review]
Looks good, just a single optional suggestion.

**Review by vuule — APPROVED** [review]

**Pull request merged**